st2common/st2common/models/utils/action_alias_utils.py

```diff
@@ -90,7 +106,16 @@ def get_extracted_param_value(self):
         param_match = r'\1["\']?(?P<\2>(?:(?<=\').+?(?=\')|(?<=").+?(?=")|{.+?}|.+?))["\']?'
         reg = re.sub(r'(\s*)' + snippets['optional'], r'(?:' + param_match + r')?', self._format)
         reg = re.sub(r'(\s*)' + snippets['required'], param_match, reg)
-        reg = '^\s*' + reg + r'\s*$'
+
+        reg_tokens = parse(reg, flags=re.DOTALL)
+
+        # Add a beginning anchor if none exists
+        if not search_regex_tokens(((AT, AT_BEGINNING), (AT, AT_BEGINNING_STRING)), reg_tokens):
```

Comment:
Just curious, how much performance overhead does those two function calls add? Did you have a chance to time it yet?

I'm wondering, because we are just working on moving all the ChatOps alias matching to the server- side which means we need to call parse for every chat message line - this means that this function should be as fast as possible otherwise user experience won't be good.

Reply:
I haven't done any performance testing yet.

Reply:
Would be good to do at least some basic ones, because as mentioned above, this code will be called for every chat message going forward.

Reply:
Here's some quick results from `line_profiler` using two "real life" regexes and strings.

## Regex 1 (worst case: does not contain beginning or ending regexes)

#### Previous code:

```
Timer unit: 1e-06 s

Total time: [scrubbed] s
File: st2common/st2common/models/utils/action_alias_utils.py
Function: get_extracted_param_value at line 34

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   ...
    78         1         2493   2493.0     29.4          ending_pairs = re.match(snippets['ending'], param_stream, re.DOTALL)
    79         1            2      2.0      0.0          has_ending_pairs = ending_pairs and ending_pairs.group(1)
    80         1            0      0.0      0.0          if has_ending_pairs:
    81                                                       kv_pairs = re.findall(snippets['pairs'], ending_pairs.group(1), re.DOTALL)
    82                                                       param_stream = param_stream.replace(ending_pairs.group(1), '')
    83         1            2      2.0      0.0          param_stream = " %s " % (param_stream)
    84                                           
    85                                                   # 2. Matching optional parameters (with default values).
    86         1         1520   1520.0     18.0          optional = re.findall(snippets['optional'], self._format, re.DOTALL)
    87                                           
    88                                                   # Transforming our format string into a regular expression,
    89                                                   # substituting {{ ... }} with regex named groups, so that param_stream
    90                                                   # matched against this expression yields a dict of params with values.
    91         1            1      1.0      0.0          param_match = r'\1["\']?(?P<\2>(?:(?<=\').+?(?=\')|(?<=").+?(?=")|{.+?}|.+?))["\']?'
    92         1         2097   2097.0     24.8          reg = re.sub(r'(\s*)' + snippets['optional'], r'(?:' + param_match + r')?', self._format)
    93         1          957    957.0     11.3          reg = re.sub(r'(\s*)' + snippets['required'], param_match, reg)

 HERE:

    94         1            1      1.0      0.0          reg = '^\s*' + reg + r'\s*$'
```

#### New code:

```
Timer unit: 1e-06 s

Total time: [scrubbed] s
File: st2common/st2common/models/utils/action_alias_utils.py
Function: get_extracted_param_value at line 50

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   ...
    94         1         1924   1924.0     21.4          ending_pairs = re.match(snippets['ending'], param_stream, re.DOTALL)
    95         1            3      3.0      0.0          has_ending_pairs = ending_pairs and ending_pairs.group(1)
    96         1            1      1.0      0.0          if has_ending_pairs:
    97                                                       kv_pairs = re.findall(snippets['pairs'], ending_pairs.group(1), re.DOTALL)
    98                                                       param_stream = param_stream.replace(ending_pairs.group(1), '')
    99         1            2      2.0      0.0          param_stream = " %s " % (param_stream)
   100                                           
   101                                                   # 2. Matching optional parameters (with default values).
   102         1         1731   1731.0     19.2          optional = re.findall(snippets['optional'], self._format, re.DOTALL)
   103                                           
   104                                                   # Transforming our format string into a regular expression,
   105                                                   # substituting {{ ... }} with regex named groups, so that param_stream
   106                                                   # matched against this expression yields a dict of params with values.
   107         1            1      1.0      0.0          param_match = r'\1["\']?(?P<\2>(?:(?<=\').+?(?=\')|(?<=").+?(?=")|{.+?}|.+?))["\']?'
   108         1         1997   1997.0     22.2          reg = re.sub(r'(\s*)' + snippets['optional'], r'(?:' + param_match + r')?', self._format)
   109         1          871    871.0      9.7          reg = re.sub(r'(\s*)' + snippets['required'], param_match, reg)
   110                                           

 HERE:

   111         1          675    675.0      7.5          reg_tokens = parse(reg, flags=re.DOTALL)
   112                                           
   113                                                   # Add a beginning anchor if none exists
   114         1          213    213.0      2.4          if not search_regex_tokens(((AT, AT_BEGINNING), (AT, AT_BEGINNING_STRING)), reg_tokens):
   115         1            2      2.0      0.0              reg = r'^\s*' + reg
   116                                           
   117                                                   # Add an ending anchor if none exists
   118         1          182    182.0      2.0          if not search_regex_tokens(((AT, AT_END), (AT, AT_END_STRING)), reg_tokens, backwards=True):
   119         1            1      1.0      0.0              reg = reg + r'\s*$'
```

According to these results, the new code adds approximately 1069µs to execution time.

## Regex 2 (more complex: contains multiple subpatterns and branches that contain beginning and ending anchors)

#### Previous code:

```
Timer unit: 1e-06 s

Total time: [scrubbed] s
File: st2common/st2common/models/utils/action_alias_utils.py
Function: get_extracted_param_value at line 34

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   ...
    78         1         2533   2533.0     28.7          ending_pairs = re.match(snippets['ending'], param_stream, re.DOTALL)
    79         1            2      2.0      0.0          has_ending_pairs = ending_pairs and ending_pairs.group(1)
    80         1            1      1.0      0.0          if has_ending_pairs:
    81                                                       kv_pairs = re.findall(snippets['pairs'], ending_pairs.group(1), re.DOTALL)
    82                                                       param_stream = param_stream.replace(ending_pairs.group(1), '')
    83         1            2      2.0      0.0          param_stream = " %s " % (param_stream)
    84                                           
    85                                                   # 2. Matching optional parameters (with default values).
    86         1         1528   1528.0     17.3          optional = re.findall(snippets['optional'], self._format, re.DOTALL)
    87                                           
    88                                                   # Transforming our format string into a regular expression,
    89                                                   # substituting {{ ... }} with regex named groups, so that param_stream
    90                                                   # matched against this expression yields a dict of params with values.
    91         1            1      1.0      0.0          param_match = r'\1["\']?(?P<\2>(?:(?<=\').+?(?=\')|(?<=").+?(?=")|{.+?}|.+?))["\']?'
    92         1         2181   2181.0     24.7          reg = re.sub(r'(\s*)' + snippets['optional'], r'(?:' + param_match + r')?', self._format)
    93         1         1194   1194.0     13.5          reg = re.sub(r'(\s*)' + snippets['required'], param_match, reg)

 HERE:

    94         1            1      1.0      0.0          reg = '^\s*' + reg + r'\s*$'
```

#### New code:

```
Timer unit: 1e-06 s

Total time: [scrubbed] s
File: st2common/st2common/models/utils/action_alias_utils.py
Function: get_extracted_param_value at line 50

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   ...
    94         1         1836   1836.0     23.6          ending_pairs = re.match(snippets['ending'], param_stream, re.DOTALL)
    95         1            3      3.0      0.0          has_ending_pairs = ending_pairs and ending_pairs.group(1)
    96         1            0      0.0      0.0          if has_ending_pairs:
    97                                                       kv_pairs = re.findall(snippets['pairs'], ending_pairs.group(1), re.DOTALL)
    98                                                       param_stream = param_stream.replace(ending_pairs.group(1), '')
    99         1            2      2.0      0.0          param_stream = " %s " % (param_stream)
   100                                           
   101                                                   # 2. Matching optional parameters (with default values).
   102         1         1518   1518.0     19.5          optional = re.findall(snippets['optional'], self._format, re.DOTALL)
   103                                           
   104                                                   # Transforming our format string into a regular expression,
   105                                                   # substituting {{ ... }} with regex named groups, so that param_stream
   106                                                   # matched against this expression yields a dict of params with values.
   107         1            1      1.0      0.0          param_match = r'\1["\']?(?P<\2>(?:(?<=\').+?(?=\')|(?<=").+?(?=")|{.+?}|.+?))["\']?'
   108         1         2018   2018.0     25.9          reg = re.sub(r'(\s*)' + snippets['optional'], r'(?:' + param_match + r')?', self._format)
   109         1          868    868.0     11.2          reg = re.sub(r'(\s*)' + snippets['required'], param_match, reg)
   110                                           

 HERE:

   111         1          491    491.0      6.3          reg_tokens = parse(reg, flags=re.DOTALL)
   112                                           
   113                                                   # Add a beginning anchor if none exists
   114         1           69     69.0      0.9          if not search_regex_tokens(((AT, AT_BEGINNING), (AT, AT_BEGINNING_STRING)), reg_tokens):
   115                                                       reg = r'^\s*' + reg
   116                                           
   117                                                   # Add an ending anchor if none exists
   118         1           53     53.0      0.7          if not search_regex_tokens(((AT, AT_END), (AT, AT_END_STRING)), reg_tokens, backwards=True):
   119                                                       reg = reg + r'\s*$'
```

According to these results, the new code adds approximately 612µs to execution time.

The total execution time (scrubbed details because they are misleading) was originally approximately 8500µs (an increase of ~7.2%).

Reply:
Thanks.

That's not too bad so we can go ahead and merge it. If it ever becomes an issue in the future, we can deal with it then :)